### ISSUE TITLE
Retain number of invalid/clipped derivatives in a JIT-compiled function?

### ISSUE THREAD

**Issue by jeffgortmaker** [google/jax#2796]
I'm using `custom_vjp` to replace some `nan` derivatives, similar to the gradient clipping example in [this tutorial](https://jax.readthedocs.io/en/latest/notebooks/Custom_derivative_rules_for_Python_code.html), but in a `jit`-compiled function. Is it possible in JAX to somehow keep track of the number of replaced/clipped values?

I've tried playing around with `value_and_grad` but couldn't figure out how to put information about the gradient into the function value.

This may just be the same as #364, but I'm hoping that there's a workaround until that gets added.

Context: I'm using JAX to estimate a model with maximum likelihood, so my objective is a sum of many log likelihoods. At some parameter values there can be overflow/underflow issues. One way to deal with this is to replace any invalid log likelihoods/derivatives with reasonable values to keep estimation going.

**Comment by mattjj** [comment]
Thanks for the question!

I think we wouldn't need quite as much as the idea in #364 because we could compute the number of nans within XLA and wouldn't need a Python callback. My understanding is this is mainly about plumbing: how do we plumb out auxiliary values from a custom vjp? So this may be related to #2574, and some effect-like things @jekbradbury has been thinking about.

Unfortunately we don't have a good solution to this plumbing problem right now. @jacobjinkelly 's clever approach in #2574 is the best idea so far, but I'm hesitant to merge that or recommend it widely because it's likely not the right long-term solution.

Let's keep thinking about this...

**Comment by jeffgortmaker** [comment]
Aha yes! That seems like pretty much exactly the same issue. Basically returning aux information with the gradient, just like it can be returned with the primal.

The fact that a (super clever/hackish) solution does exist makes me optimistic. Thanks for taking a look.

**Comment by jeffgortmaker** [comment]
I'm coming back to JAX after a few years and found that it now supports plenty of callbacks! For anyone else with this issue, here's some example code that I got working.

The idea is that there's some downstream `f` function that can sometimes produce invalid derivatives. Here I've chosen the square root function, which produces an invalid gradient at zero.

The first chunk of code defines a custom JVP wrapper around `f`, which replaces bad derivatives with zeros and logs these. The second chunk of code defines a custom VJP wrapper around `f`, which does something similar but in the backward pass.

```python
import jax
import jax.numpy as jnp

def f(x):
    return jnp.sqrt(x)

def log(bad):
    log.count += bad

@jax.custom_jvp
def g(x):
    return f(x)

@g.defjvp
def g_jvp(x, g):
    x, g = jax.jvp(f, x, g)
    bad = ~jnp.isfinite(g)
    g = jnp.where(bad, 0, g)
    jax.debug.callback(log, bad)
    return x, g

log.count = 0
print(jax.jit(jax.jacfwd(g))(2.))
print(log.count)
print(jax.jit(jax.jacfwd(g))(0.))
print(log.count)
print()

@jax.custom_vjp
def h(x):
    return f(x)

def h_fwd(x):
    return jax.vjp(f, x)

def h_bwd(v, g):
    vg, = v(g)
    bad = ~jnp.isfinite(vg)
    vg = jnp.where(bad, 0, vg)
    jax.debug.callback(log, bad)
    return vg,

h.defvjp(h_fwd, h_bwd)

log.count = 0
print(jax.jit(jax.jacrev(h))(2.))
print(log.count)
print(jax.jit(jax.jacrev(h))(0.))
print(log.count)
```

This gives:

```
0.35355335
0
0.0
1

0.35355335
0
0.0
1
```

Hope this helps anyone else who tries to do something like this. And hopefully I'm not doing something too terrible here -- seems like a reasonable use of `debug.callback` but I'm not an expert.
